### PR TITLE
Fix token up to mint and add a flag to enable/disable mint

### PR DESCRIPTION
**Pull request by 0xaptosj**
- Add new view function to return user mint balance
- Store mint balance (i.e. how many token left to mint) instead of how many user has minted in the tracker, this is to keep consistency with token minter contract because token minter tracks balance left to mint instead of minted so far.
- Add a flag to config that can disable / enable mint

Screenshot 2024-08-22 at 3 34 58 PM

Screenshot 2024-08-22 at 3 32 56 PM

### CODE REVIEW

File: templates/nft-minting-dapp-template/frontend/hooks/useGetCollectionData.ts [hunk 1]
```diff
@@ -112,12 +138,6 @@ export function useGetCollectionData(collection_id: string = config.collection_i
                   cdn_image_uri
                 }
 							}
-							current_collection_ownership_v2_view(
-								where: { collection_id: { _eq: $collection_id } }
-								order_by: { last_transaction_version: desc }
-							) {
-								owner_address
```
Comment (0xaptosj) [inline templates/nft-minting-dapp-template/frontend/hooks/useGetCollectionData.ts:119]: we are not using this

File: templates/nft-minting-dapp-template/frontend/hooks/useGetCollectionData.ts [hunk 2]
```diff
@@ -79,15 +75,45 @@ async function getStartAndEndTime(collection_id: string): Promise<[start: Date,
   ];
 }
 
+async function getUserMintBalance(
```
Comment (0xmaayan) [inline templates/nft-minting-dapp-template/frontend/hooks/useGetCollectionData.ts:78]: this should live in the `view-functions` folder https://github.com/aptos-labs/create-aptos-dapp/tree/main/templates/nft-minting-dapp-template/frontend/view-functions

File: templates/nft-minting-dapp-template/frontend/hooks/useGetCollectionData.ts [hunk 3]
```diff
@@ -79,15 +75,45 @@ async function getStartAndEndTime(collection_id: string): Promise<[start: Date,
...
+
   return useQuery({
     queryKey: ["app-state", collection_id],
     refetchInterval: 1000 * 30,
     queryFn: async () => {
       try {
         if (!collection_id) return null;
 
-        const [startDate, endDate, isMintInfinite] = await getStartAndEndTime(collection_id);
+        const mintStageRes = await aptosClient().view<[{ vec: [string] | [] }]>({
```
Comment (0xmaayan) [inline templates/nft-minting-dapp-template/frontend/hooks/useGetCollectionData.ts:103]: this is also

File: templates/nft-minting-dapp-template/frontend/hooks/useGetCollectionData.ts [hunk 4]
```diff
@@ -79,15 +75,45 @@ async function getStartAndEndTime(collection_id: string): Promise<[start: Date,
...
+          payload: {
+            function: `${AccountAddress.from(MODULE_ADDRESS)}::launchpad::get_active_or_next_mint_stage`,
+            functionArguments: [collection_id],
+          },
+        });
+
+        if (mintStageRes[0].vec.length === 0) return null;
```
Comment (0xmaayan) [inline templates/nft-minting-dapp-template/frontend/hooks/useGetCollectionData.ts:110]: a bit confused - is this a check on whether there is an active mint, i.e a collection has been made, and if there is isn't we wont do the indexer call?

File: templates/token-minting-dapp-template/frontend/hooks/useGetAssetData.ts [hunk 5]
```diff
@@ -36,20 +36,20 @@ interface MintData {
...
   yourBalance: number;
-  totalAbleToMint: number;
+  userMintBalance: number;
   asset: FungibleAsset;
   isMintActive: boolean;
 }
 
-async function getMintLimit(fa_address: string): Promise<number> {
-  const mintLimitRes = await aptosClient().view<[{ vec: [string] }]>({
+async function getUserMintBalance(user_address: string, fa_address: string): Promise<number> {
```
Comment (0xmaayan) [inline templates/token-minting-dapp-template/frontend/hooks/useGetAssetData.ts:44]: this should live in the view-functions folder

File: templates/token-minting-dapp-template/move/sources/launchpad.move [hunk 6]
```diff
@@ -313,16 +313,17 @@ module launchpad_addr::launchpad {
     }
 
     #[view]
-    /// Get current minted amount by an address
-    public fun get_current_minted_amount(
+    /// Get mint balance, i.e. how many tokens user can mint
+    /// e.g. If the mint limit is 1, user has already minted 1, balance is 0
+    public fun get_mint_balance(
```
Comment (0xmaayan) [inline templates/token-minting-dapp-template/move/sources/launchpad.move:318]: @GotenJBZ could you help with a review?

File: templates/nft-minting-dapp-template/move/sources/launchpad.move [hunk 7]
```diff
@@ -380,6 +380,18 @@ module launchpad_addr::launchpad {
         amount * fee
     }
 
+    #[view]
+    /// Get mint balance for the stage, i.e. how many NFT user can mint
+    /// e.g. If the mint limit is 1, user has already minted 1, balance is 0
+    public fun get_mint_balance(collection_obj: Object<Collection>, stage_name: String, user_addr: address): u64 {
```
Comment (0xmaayan) [inline templates/nft-minting-dapp-template/move/sources/launchpad.move:386]: @GotenJBZ could you help with a review?

File: templates/nft-minting-dapp-template/frontend/hooks/useGetCollectionData.ts (hunk 4 above)
Comment (0xaptosj) [reply templates/nft-minting-dapp-template/frontend/hooks/useGetCollectionData.ts]: updated, previous code will throw error when mint stage not found, now ill only return collection data if mint stage not found

**Review by 0xmaayan — COMMENTED** [review]
could you update CHANGELOG

**Comment by 0xaptosj** [comment]
> could you update CHANGELOG

updated

**Review by GotenJBZ — APPROVED** [review]

**Review by 0xmaayan — APPROVED** [review]

**Pull request merged**